File: PersistenceService/v970426/lib/ObjectStore/HashedStore.cc
/*
 * Copyright (C) 1993
 *
 * Department of Computing Science,
 * The University,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: HashedStore.cc,v 1.10 1995/07/03 14:29:27 ngdp Exp $
 */

#ifndef STRSTREAM_H_
#  include <os/strstream.h>
#endif

#ifndef PERSISTENCE_SERVICE

#ifndef MEMORY_H_
#  include <os/memory.h>
#endif

#endif

#ifndef STRING_H_
#  include <os/string.h>
#endif

#ifndef SYS_PARAM_H_
#  include <os/sys/param.h>
#endif

#ifndef SYS_STAT_H_
#  include <os/sys/stat.h>
#endif

#ifndef DIRENT_H_
#  include <os/dirent.h>
#endif

#ifndef OBJSTATE_H_
#  include <Common/ObjState.h>
#endif

#ifndef OBJSTORE_H_
#  include <ObjectStore/ObjStore.h>
#endif

#ifndef HASHEDSTORE_H_
#  include <ObjectStore/HashedStore.h>
#endif

#ifndef STORELIST_H_
#  include <ObjectStore/StoreList.h>
#endif

static osCreatorFunc createHSInstance;

#ifdef STRUCT_INIT_BUG
static SetUpInfo LHSetup = { 0, 0, 0, &createHSInstance, HashedStoreType, 0 };
#else
static SetUpInfo LHSetup = { FALSE, FALSE, 0, &createHSInstance, HashedStoreType, 0 };
#endif

static const NUMBEROFDIRECTORIES = 255;


ObjectStore* createHSInstance ( const char* locationOfStore )
{
    return new HashedStore(locationOfStore);
}

HashedStoreSetUp::HashedStoreSetUp ()
				   : FragmentedStoreSetUp(LHSetup)
{
#ifdef DEBUG
    debug_stream.lock();
    
    debug_stream << CONSTRUCTORS << FAC_OBJECT_STORE << VIS_PUBLIC;
    debug_stream << "HashedStoreSetUp::HashedStoreSetUp ()" << endl;
    debug_stream.unlock();
    
#endif
  
#ifdef GCC_STATIC_INIT_BUG
    if (!LHSetup._setUp)
    {
	LHSetup._new = &createHSInstance;
	LHSetup._tn = HashedStoreType;
	LHSetup._setUp = TRUE;
	ObjectStore::addToList(&LHSetup);
    }
#endif   
}

HashedStoreSetUp::~HashedStoreSetUp () {}

/*
 * Public constructors and destructor
 */

HashedStore::HashedStore ()
			 : FragmentedStore()
{
#ifdef DEBUG
    debug_stream.lock();
    
    debug_stream << CONSTRUCTORS << FAC_OBJECT_STORE << VIS_PUBLIC;
    debug_stream << "HashedStore::HashedStore()" << endl;
    debug_stream.unlock();
    
#endif
}

HashedStore::HashedStore ( const char* locationOfStore )
			 : FragmentedStore()
{
#ifdef DEBUG
    debug_stream.lock();
    
    debug_stream << CONSTRUCTORS << FAC_OBJECT_STORE << VIS_PUBLIC;
    debug_stream << "HashedStore::HashedStore (\""
		 << locationOfStore << "\")" << endl;
    debug_stream.unlock();
    
#endif
    setupStore(locationOfStore);
}

HashedStore::~HashedStore ()
{
#ifdef DEBUG
    debug_stream.lock();
    
    debug_stream << DESTRUCTORS << FAC_OBJECT_STORE << VIS_PUBLIC;
    debug_stream << "HashedStore::~HashedStore()" << endl;
    debug_stream.unlock();
    
#endif
}

/*
 * Given a type name return an objectstate that contains all of the
 * uids of objects of that type
 */

Boolean HashedStore::allObjUids ( const TypeName tName,
				  ObjectState& state,
				  ObjectStore::StateStatus match )
{
    char *storeName = locateStore(getStoreName());
    DIR *rdir, *dir = 0;
    char directory[MAXPATHLEN+1];
    char *cPtr = tName;
    int slen = 0;

    /* storeName ALWAYS has a trailing '/' */
	
    ::strcpy(directory, storeName);
	
    /* Does typename start with a '/' if so skip over */
    
    if (tName[0] == '/')
	cPtr++;
    
    ::strcat(directory, cPtr);

    slen = ::strlen(directory);

    if (directory[slen-1] != '/')
    {
	slen++;
	::strcat(directory, "/");
    }

    cPtr = directory + slen;
    
    rdir = ::opendir(directory);
    
    if (rdir)
    {
	dirent *entry;
	struct stat statbuf;

	while ((entry = ::readdir(rdir)))
	{
	    /* Only accept dirs with names surrounded by <> */
	    
	    if ((entry->d_name[0] == '<') &&
		(entry->d_name[::strlen(entry->d_name)-1] == '>'))
	    {
		::strcpy(cPtr,entry->d_name);
	    
		if ((::stat(directory, &statbuf) == 0) &&
		    S_ISDIR(statbuf.st_mode) &&
		    (dir = ::opendir(directory)))
		{
		    dirent *subEntry;
		    char *tPtr = directory + ::strlen(directory);

		    while ((subEntry = ::readdir(dir)))
		    {
			::strcpy(tPtr,"/");
			::strcat(directory,subEntry->d_name);
		    
			if ((::stat(directory, &statbuf) == 0) &&
			    S_ISREG(statbuf.st_mode))
			{
			    Uid aUid(subEntry->d_name, TRUE);
			    
			    if ((aUid != NIL_UID) &&
				((match == ObjectStore::OS_UNKNOWN) ||
				 (isType(aUid, tName, match))))
				aUid.pack(state);
			}
		    }
		}
	    
		::closedir(dir);
	    }
	}
    
	::closedir(rdir);
    }
    
    /* terminate list */

    NIL_UID.pack(state);

    if (storeName)
#ifdef GCC_ARRAY_DELETE_BUG
	delete storeName;
#else
    	delete [] storeName;
#endif

    return TRUE;
	
}

/*
 * Function to return the file name for the state of the object
 * identified by the Uid and TypeName. 
 * Return char * is dynamically allocated and must be freed by caller
 */

char* HashedStore::genPathName ( const Uid& objUid, 
				 const TypeName tName,
				 ObjectStore::StateType ) const
{
    char *storeName = locateStore(getStoreName());
    char uidBuff[MAXPATHLEN+1];
    ostrstream os(uidBuff, MAXPATHLEN+1, ios::out);
    char *fname = 0;
    char *cPtr = tName;
    unsigned long uidHash = objUid.hash();
    
    os << "<" << uidHash % NUMBEROFDIRECTORIES << ">/";
    os << objUid << ends;
    
    fname = new char[strlen(storeName) + strlen(tName) + strlen(uidBuff) + 5];

    ::strcpy(fname, storeName);

    if (tName[0] == '/')
	cPtr++;
    
    ::strcat(fname, cPtr);

    if (fname[::strlen(fname)-1] != '/')
	::strcat(fname, "/");

    ::strcat(fname, uidBuff);

    if (storeName)
#ifdef GCC_ARRAY_DELETE_BUG
	delete storeName;
#else
    	delete [] storeName;
#endif
    
    return fname;
}

Boolean HashedStore::supressEntry ( const char *name ) const
{
    if ((name[0] == '<') &&
	(name[::strlen(name)-1] == '>'))
	return TRUE;
    else
	return FragmentedStore::supressEntry(name);
}

#ifdef NO_INLINES
#  define HASHEDSTORE_CC_
#  include "ObjectStore/HashedStore.n"
#  undef HASHEDSTORE_CC_
#endif

File: JTSArjuna/JTSArjuna-2.0_Alpha_2_1/ArjunaLite/classes/com/arjuna/ArjunaLite/Implementation/ObjectStore/HashedStore.java
/*
 * Copyright (C) 1998, 1999, 2000,
 *
 * Arjuna Solutions Limited,
 * Newcastle upon Tyne,
 * Tyne and Wear,
 * UK.  
 *
 * $Id: HashedStore.java,v 1.1 2000/02/25 14:04:09 nmcl Exp $
 */

package com.arjuna.ArjunaLite.Implementation.ObjectStore;

import com.arjuna.ArjunaCommon.Common.*;
import com.arjuna.ArjunaLite.ArjunaLiteNames;
import com.arjuna.ArjunaLite.Common.ArjunaLiteEnvironment;
import com.arjuna.ArjunaLite.Atomic.*;
import com.arjuna.ArjunaCommon.Common.ClassName;
import com.arjuna.ArjunaLite.Interface.ObjectStore;
import java.io.*;
import java.io.File;

import com.arjuna.ArjunaLite.Common.ObjectStoreException;
import java.io.IOException;
import java.io.FileNotFoundException;
import java.lang.NumberFormatException;

/*
 * Should be derived from FragmentedStore, but we currently
 * don't have such an implementation in Java.
 */

public class HashedStore extends ShadowNoFileLockStore
{

public int typeIs ()
    {
	return ObjectStoreType.HASHED;
    }

public ClassName className ()
    {
	return ArjunaLiteNames.Implementation_ObjectStore_HashedStore();
    }

public static ClassName name ()
    {
	return ArjunaLiteNames.Implementation_ObjectStore_HashedStore();
    }

    /*
     * Have to return as a ShadowingStore because of
     * inheritence.
     */
    
public static ShadowingStore create ()
    {
	return new HashedStore("");
    }

public static ShadowingStore create (Object[] param)
    {
	if (param == null)
	    return null;

	String location = (String) param[0];

	return new HashedStore(location);
    }

/*
 * Given a type name return an objectstate that contains all of the
 * uids of objects of that type
 */

public boolean allObjUids (String tName, InputObjectState state, int match) throws ObjectStoreException
    {
	/*
	 * Directory ALWAYS has a trailing '/'
	 */
	
	String directory = locateStore(getStoreName());
	OutputObjectState store = new OutputObjectState();
	
	/* Does typename start with a '/' if so skip over */

	if ((tName != null) && (tName.charAt(0) == File.separatorChar))
	    directory = directory + tName.substring(1, tName.length());
	else
	    directory = directory + tName;

	if (!directory.endsWith("/"))
	    directory = directory + "/";

	File f = new File(directory);
	String[] entry = f.list();

	if ((entry != null) && (entry.length > 0))
	{
	    for (int i = 0; i < entry.length; i++)
	    {
		File dir = new File(entry[i]);

		if (dir.isDirectory())
		{
		    String[] dirEnt = dir.list();

		    for (int j = 0; j < dirEnt.length; j++)
		    {
			try
			{
			    Uid aUid = new Uid(entry[i]);

			    if ((aUid.notEquals(Uid.nullUid())) && ((match == ObjectStore.OS_UNKNOWN)) ||
				(isType(aUid, tName, match)))
			    {
				aUid.pack(store);
			    }
			}
			catch (NumberFormatException e)
			{
			    /*
			     * Not a number at start of file.
			     */
			}
			catch (IOException e)
			{
			    throw new ObjectStoreException("HashedStore.allObjUids - could not pack Uid.");
			}
		    }
		}
	    }
	}
	
	/* terminate list */

	try
	{
	    Uid.nullUid().pack(store);
	}
	catch (IOException e)
	{
	    throw new ObjectStoreException("HashedStore.allObjUids - could not pack end of list Uid.");
	}

	state.setBuffer(store.buffer());

	store = null;
	
	return true;
    }

    /*
     * Protected constructors and destructor
     */

protected HashedStore ()
    {
	super();
    }

protected HashedStore (String locationOfStore)
    {
	super();

	try
	{
	    setupStore(locationOfStore);
	}
	catch (ObjectStoreException e)
	{
	    System.out.println(e.getMessage());
	    System.exit(0);
	}
    }
    
    /*
     * Function to return the file name for the state of the object
     * identified by the Uid and TypeName. 
     * Return char * is dynamically allocated and must be freed by caller
     */

protected String genPathName (Uid objUid, String tName, int m) throws ObjectStoreException
    {
	String storeName = locateStore(getStoreName());
	String fname = null;
	String cPtr = null;
	int uidHash = objUid.hashCode();
	String uidString = objUid.toString();
	String hashDir = uidHash % HashedStore.NUMBEROFDIRECTORIES + "/" + uidString;

	/*
	 * Now convert all ':' to '_' to be portable across file systems.
	 */

	String os = uidString.replace(':', '_');

	uidString = null;
	
	if ((tName == null) || (tName.length() == 0))
	    cPtr = "";
	else
	{
	    cPtr = tName;

	    /*
	     * Convert Unix separators to 'other', i.e., Windows!
	     */

	    if (FileSystemStore.rewriteSeparator && (cPtr.indexOf(FileSystemStore.unixSeparator) != -1))
		cPtr = cPtr.replace(FileSystemStore.unixSeparator, File.separatorChar);
	}
	
	/*
	 * storeName always ends in '/' so we can remove any
	 * at the start of the type name.
	 */
	
	if (cPtr.charAt(0) == File.separatorChar)
	    cPtr = cPtr.substring(1, cPtr.length());

	if (cPtr.charAt(cPtr.length() -1) != File.separatorChar)
	    fname = storeName + cPtr + File.separator + hashDir;
	else
	    fname = storeName + cPtr + hashDir;
	
	/*
	 * Make sure we don't end in a '/'.
	 */
	
	if (fname.charAt(fname.length() -1) == File.separatorChar)
	    fname = fname.substring(0, fname.length() -2);

	return fname;
    }

private static final int DEFAULT_NUMBER_DIRECTORIES = 255;
private static int NUMBEROFDIRECTORIES = DEFAULT_NUMBER_DIRECTORIES;

    static
	{
	    String numberOfDirs = PropertyManager.getProperty(ArjunaLiteEnvironment.HASHED_DIRECTORIES);

	    if (numberOfDirs != null)
	    {
		try
		{
		    Integer i = new Integer(numberOfDirs);

		    NUMBEROFDIRECTORIES = i.intValue();

		    if (NUMBEROFDIRECTORIES <= 0)
		    {
			System.err.println("WARNING - HashedStore - invalid number of hash directories: "+numberOfDirs);
			System.err.println("WARNING - Will use default.");

			NUMBEROFDIRECTORIES = DEFAULT_NUMBER_DIRECTORIES;
		    }
		}
		catch (NumberFormatException e)
		{
		    System.err.println("HashedStore - invalid number of hash directories: "+numberOfDirs);
		    System.exit(0);
		}
		catch (Exception e)
		{
		    System.err.println("HashedStore caught exception: "+e);
		    System.exit(0);
		}
	    }
	}
    
};
